packages/multi-test/src/app.rs

```diff
@@ -2580,4 +2580,129 @@ mod test {
+
+            let error_code_id = app.store_code(error::contract(true));
+            let caller_code_id = app.store_code(caller::contract());
+
+            // set up contracts
+            let msg = EmptyMsg {};
+            let caller_addr = app
+                .instantiate_contract(caller_code_id, owner.clone(), &msg, &[], "caller", None)
+                .unwrap();
+            let error_addr = app
+                .instantiate_contract(error_code_id, owner, &msg, &[], "error", None)
+                .unwrap();
+
+            // execute should error
+            let msg = WasmMsg::Execute {
+                contract_addr: error_addr.into(),
+                msg: to_binary(&EmptyMsg {}).unwrap(),
+                funds: vec![],
+            };
+            let err = app
+                .execute_contract(Addr::unchecked("random"), caller_addr, &msg, &[])
+                .unwrap_err();
+
+            // we can downcast to get the original error
+            let source: &StdError = err.downcast_ref().unwrap();
+            if let StdError::GenericErr { msg } = source {
+                assert_eq!(msg, "Handle failed");
+            } else {
+                panic!("wrong StdError variant");
+            }
+
+            // we're expecting exactly 4 nested error types
+            // (the original error, execute msg context, 2 WasmMsg contexts)
+            assert_eq!(err.chain().count(), 4);
+        }
+
+        #[test]
+        fn double_nested_call() {
+            let owner = Addr::unchecked("owner");
+            let mut app = App::default();
+
+            let error_code_id = app.store_code(error::contract(true));
+            let caller_code_id = app.store_code(caller::contract());
+
+            // set up contracts
+            let msg = EmptyMsg {};
+            let caller_addr1 = app
+                .instantiate_contract(caller_code_id, owner.clone(), &msg, &[], "caller", None)
+                .unwrap();
+            let caller_addr2 = app
+                .instantiate_contract(caller_code_id, owner.clone(), &msg, &[], "caller", None)
+                .unwrap();
+            let error_addr = app
+                .instantiate_contract(error_code_id, owner, &msg, &[], "error", None)
+                .unwrap();
+
+            // caller1 calls caller2, caller2 calls error
+            let msg = WasmMsg::Execute {
+                contract_addr: caller_addr2.into(),
+                msg: to_binary(&WasmMsg::Execute {
+                    contract_addr: error_addr.into(),
+                    msg: to_binary(&EmptyMsg {}).unwrap(),
+                    funds: vec![],
+                })
+                .unwrap(),
+                funds: vec![],
+            };
+            let err = app
+                .execute_contract(Addr::unchecked("random"), caller_addr1, &msg, &[])
+                .unwrap_err();
+
+            // uncomment to have the test fail and see how the error stringifies
+            // panic!("{:?}", err);
```

Comment:
A dirty way to take a peek at the debug formatted error with two levels of nesting.

Reply:
I get:

```
---- app::test::errors::double_nested_call stdout ----
thread 'app::test::errors::double_nested_call' panicked at 'error executing WasmMsg:
sender: random
Execute { contract_addr: "Contract #0", msg: Binary([123, 34, 101, 120, 101, 99, 117, 116, 101, 34, 58, 123, 34, 99, 111, 110, 116, 114, 97, 99, 116, 95, 97, 100, 100, 114, 34, 58, 34, 67, 111, 110, 116, 114, 97, 99, 116, 32, 35, 49, 34, 44, 34, 109, 115, 103, 34, 58, 34, 101, 121, 74, 108, 101, 71, 86, 106, 100, 88, 82, 108, 73, 106, 112, 55, 73, 109, 78, 118, 98, 110, 82, 121, 89, 87, 78, 48, 88, 50, 70, 107, 90, 72, 73, 105, 79, 105, 74, 68, 98, 50, 53, 48, 99, 109, 70, 106, 100, 67, 65, 106, 77, 105, 73, 115, 73, 109, 49, 122, 90, 121, 73, 54, 73, 109, 85, 122, 77, 68, 48, 105, 76, 67, 74, 109, 100, 87, 53, 107, 99, 121, 73, 54, 87, 49, 49, 57, 102, 81, 61, 61, 34, 44, 34, 102, 117, 110, 100, 115, 34, 58, 91, 93, 125, 125]), funds: [] }

Caused by:
    0: error executing WasmMsg:
       sender: Contract #0
       Execute { contract_addr: "Contract #1", msg: Binary([123, 34, 101, 120, 101, 99, 117, 116, 101, 34, 58, 123, 34, 99, 111, 110, 116, 114, 97, 99, 116, 95, 97, 100, 100, 114, 34, 58, 34, 67, 111, 110, 116, 114, 97, 99, 116, 32, 35, 50, 34, 44, 34, 109, 115, 103, 34, 58, 34, 101, 51, 48, 61, 34, 44, 34, 102, 117, 110, 100, 115, 34, 58, 91, 93, 125, 125]), funds: [] }
    1: error executing WasmMsg:
       sender: Contract #1
       Execute { contract_addr: "Contract #2", msg: Binary([123, 125]), funds: [] }
    2: Contract returned an error on execute msg:
       EmptyMsg
    3: Generic error: Handle failed', packages/multi-test/src/app.rs:2693:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
This is a great output.
I just also realise we need a better debug encoding for Binary.

Added https://github.com/CosmWasm/cosmwasm/issues/1199 which could be an interesting follow-up sometime.

Reply:
Sounds good! It's probably not super important that the binary stuff is readable here - it's usually going to be something predictable like `WasmMsg::Execute`, and then the contract execute msg itself will be shown in its deserialized form. Like `2:` in your output.